From a6dcce83a210167b2c023eb208609e909186cdd8 Mon Sep 17 00:00:00 2001
From: Antonio Terceiro <terceiro@debian.org>
Date: Thu, 12 Jan 2023 09:43:33 -0300
Subject: [PATCH] Query dpkg for supported architectures

Instead of maintaining a hardcoded list, just query dpkg-architecture
for all supported architectures, and allow those. This way people
working on new ports and using apt-cacher have one less place to patch.

To avoid a giant list, and a giant configuration file line, drop the
architectures that have dashes (-) in their names, so we avoid more
exotic ports like uclinux-*, aix-* etc, keeping only the traditional
Linux + glibc ports 99.9999% of us use.
---
 config/apt-cacher.conf |  2 +-
 lib/apt-cacher.pl      | 31 +------------------------------
 2 files changed, 2 insertions(+), 31 deletions(-)

diff --git a/config/apt-cacher.conf b/config/apt-cacher.conf
index e6c95ef..1cf41f0 100644
--- a/config/apt-cacher.conf
+++ b/config/apt-cacher.conf
@@ -187,7 +187,7 @@ user = www-data
 # (see below).
 #
 #supported_archs = i386, amd64
-#supported_archs = avr32, amd64, alpha, arm, arm64, armel, armhf, hppa, hurd-i386, i386, ia64, kfreebsd-amd64, kfreebsd-i386, m32r, m68k, mips, mipsel, netbsd-alpha, netbsd-i386, powerpc, powerpcspe, ppc64, s390, s390x, sh4, sparc, sparc64, x32
+#supported_archs = armhf, armel, mipsn32, mipsn32el, mipsn32r6, mipsn32r6el, mips64, mips64el, mips64r6, mips64r6el, powerpcspe, x32, arm64ilp32, alpha, amd64, arc, armeb, arm, arm64, avr32, hppa, loong64, i386, ia64, m32r, m68k, mips, mipsel, mipsr6, mipsr6el, nios2, or1k, powerpc, powerpcel, ppc64, ppc64el, riscv64, s390, s390x, sh3, sh3eb, sh4, sh4eb, sparc, sparc64, tilegx
 
 # List of Ubuntu release names used to expand %VALID_UBUNTU_RELEASE_NAMES% in
 # *_files_regexp (see below). This is required to allow the Ubuntu installer to
diff --git a/lib/apt-cacher.pl b/lib/apt-cacher.pl
index ee14952..3002532 100755
--- a/lib/apt-cacher.pl
+++ b/lib/apt-cacher.pl
@@ -60,36 +60,7 @@ sub read_config {
 		  request_timeout => 30,
 		  return_buffer_size => 1048576, # 1Mb
 		  reverse_path_map => 1,
-		  supported_archs => join(', ', qw(
-						      avr32
-						      amd64
-						      alpha
-						      arm
-						      arm64
-						      armel
-						      armhf
-						      hppa
-						      hurd-i386
-						      i386
-						      ia64
-						      kfreebsd-amd64
-						      kfreebsd-i386
-						      m32r
-						      m68k
-						      mips
-						      mipsel
-						      netbsd-alpha
-						      netbsd-i386
-						      powerpc
-						      powerpcspe
-						      ppc64
-						      s390
-						      s390x
-						      sh4
-						      sparc
-						      sparc64
-						      x32
-						 )),
+		  supported_archs => join(', ', grep { $_ !~ /-/ } Dpkg::Arch::get_valid_arches()),
 		  ubuntu_release_names => join(', ', get_ubuntu_names()),
 		  user => $>,
 
-- 
2.39.0

